experimental/virtio/src/queue.rs

```diff
@@ -0,0 +1,606 @@
+#[repr(C, align(16))]
+#[derive(Debug)]
+struct Desc {
+    /// The guest-physical address of the buffer.
+    ///
+    /// We use an identity mapping, so it is also the guest virtual address.
+    addr: u64,
+    /// The lengths of the buffer.
+    length: u32,
+    /// Flags providing more info about this descriptor.
+    flags: DescFlags,
+    /// The index of the next descriptor in the chain if this is not the end (i.e. flags contain
+    /// `DescFlags::VIRTQ_DESC_F_NEXT`), ignored otherwise.
+    ///
+    /// Not used for now, as we don't currently support chaining.
+    next: u16,
+}
+
+impl Desc {
+    fn new(flags: DescFlags, addr: u64, length: u32) -> Self {
+        assert!(
+            !flags.contains(DescFlags::VIRTQ_DESC_F_INDIRECT),
+            "Indirect descriptors not supported."
+        );
+        assert!(
+            !flags.contains(DescFlags::VIRTQ_DESC_F_NEXT),
+            "Chained descriptors not supported."
+        );
+        Self {
+            addr,
+            length,
+            flags,
+            // For simplicity we don't set up descriptor chains.
+            next: 0,
+        }
+    }
+}
+
+bitflags! {
+    /// Flags about the available and used rings.
+    struct RingFlags: u16 {
+        /// This indicates that the owner of the ring does not require queue notifications.
+        const NO_NOTIFY = 1;
+    }
+}
+
+/// The ring buffer that indicates which descriptors have been made available by the driver.
+///
+/// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-380006>.
+#[repr(C, align(2))]
+struct AvailRing<const QUEUE_SIZE: usize> {
+    /// Driver-specific flags for the queue.
+    flags: RingFlags,
+    /// The next index that will be used in the ring (modulo `QUEUE_SIZE`).
+    ///
+    /// Starts at 0 and always increments. Must never be decremented.
+    idx: u16,
+    /// The ring-buffer containing indices of the heads of available descriptor chains.
+    ring: [u16; QUEUE_SIZE],
+}
+
+impl<const QUEUE_SIZE: usize> Default for AvailRing<QUEUE_SIZE> {
+    fn default() -> Self {
+        Self {
+            // We implement all drivers via polling, so don't want notifications.
+            flags: RingFlags::NO_NOTIFY,
+            idx: 0,
+            ring: [0; QUEUE_SIZE],
+        }
+    }
+}
+
+/// The ring buffer that indicates which available descriptors have been consumed by the device.
+///
+/// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-430008>.
+#[repr(C, align(4))]
+struct UsedRing<const QUEUE_SIZE: usize> {
+    /// Device-specific flags for the queue.
+    flags: RingFlags,
+    /// The next index that will be used in the ring (modulo `QUEUE_SIZE`).
+    ///
+    /// Starts at 0 and always increments. Must never be decremented.
+    idx: u16,
+    /// The ring-buffer containing the used elements.
+    ring: [UsedElem; QUEUE_SIZE],
+}
+
+impl<const QUEUE_SIZE: usize> Default for UsedRing<QUEUE_SIZE> {
```

Comment:
As I understand it this pretty much zero-initializes the data structure (no flags, and all `UsedElem`s are empty). Would `#[derive(Default)]` not work on `UsedRing`?

Reply:
Unfortunatley not, as Default is not implemented for const generically sized arrays, only array with fixed sizes up to 32. So I have to manually implement it for any structs that have arrays with sizes based on const generics.